### ISSUE TITLE
Arango 3.4 Search Bug (context leaking?)

### ISSUE THREAD

**Issue by aveiros** [arangodb/arangodb#8213]
## My Environment

* __ArangoDB Version__:        3.4.1, 3.4.2, 3.4.2-1
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:    ArangoDB Starter in Docker
* __Infrastructure__:               own
* __Operating System__:        MacOS 10.13.4
* __Total RAM in your machine__:        32Gb
* __Disks in use__:        SSD
* __Used Package__:              Docker - official Docker library

I have been testing SEARCH on the following graph data:

```
(v:person1) <---- e:married -----> (v:person2)
            <---- e:married -----> (v:person3)
```

```
(v:person4) <---- e:married -----> (v:person5)
```

// entities collection (vertex collection)
```
[
  {
    "_key": "person1",
    "_id": "entities/person1",
    "_rev": "_YOr40eu--_",
    "type": "person",
    "id": "person1"
  },
  {
    "_key": "person5",
    "_id": "entities/person5",
    "_rev": "_YOr48rO---",
    "type": "person",
    "id": "person5"
  },
  {
    "_key": "person4",
    "_id": "entities/person4",
    "_rev": "_YOr5IGu--_",
    "type": "person",
    "id": "person4"
  },
  {
    "_key": "person3",
    "_id": "entities/person3",
    "_rev": "_YOr5PBK--_",
    "type": "person",
    "id": "person3"
  },
  {
    "_key": "person2",
    "_id": "entities/person2",
    "_rev": "_YOr5Umq--_",
    "type": "person",
    "id": "person2"
  }
]
```

// links collection (edge collection)
```
[
  {
    "_key": "3301",
    "_id": "links/3301",
    "_from": "entities/person1",
    "_to": "entities/person2",
    "_rev": "_YOrbp_S--_",
    "type": "relationship",
    "subType": "married",
    "from": "person1",
    "to": "person2"
  },
  {
    "_key": "3377",
    "_id": "links/3377",
    "_from": "entities/person4",
    "_to": "entities/person5",
    "_rev": "_YOrbxN2--_",
    "type": "relationship",
    "subType": "married",
    "from": "person4",
    "to": "person5"
  },
  {
    "_key": "3346",
    "_id": "links/3346",
    "_from": "entities/person1",
    "_to": "entities/person3",
    "_rev": "_YOrb4kq--_",
    "type": "relationship",
    "subType": "married",
    "from": "person1",
    "to": "person3"
  }
]
```

entities_view
```
{
  "writebufferSizeMax": 33554432,
  "consolidationPolicy": {
    "type": "bytes_accum",
    "threshold": 0.10000000149011612
  },
  "globallyUniqueId": "hB4A95C21732A/218",
  "id": "218",
  "writebufferActive": 0,
  "consolidationIntervalMsec": 60000,
  "cleanupIntervalStep": 10,
  "links": {
    "entities": {
      "analyzers": [
        "identity"
      ],
      "fields": {},
      "includeAllFields": true,
      "storeValues": "id",
      "trackListPositions": false
    }
  },
  "type": "arangosearch",
  "writebufferIdle": 64
}
```

links_view
```
{
  "writebufferSizeMax": 33554432,
  "consolidationPolicy": {
    "type": "bytes_accum",
    "threshold": 0.10000000149011612
  },
  "globallyUniqueId": "hB4A95C21732A/181",
  "id": "181",
  "writebufferActive": 0,
  "consolidationIntervalMsec": 60000,
  "cleanupIntervalStep": 10,
  "links": {
    "links": {
      "analyzers": [
        "identity"
      ],
      "fields": {},
      "includeAllFields": true,
      "storeValues": "id",
      "trackListPositions": false
    }
  },
  "type": "arangosearch",
  "writebufferIdle": 64
}
```

I'm running the following query
```
// person
FOR org IN entities_view
SEARCH org.type == "person"

        // married
        LET marriedIds = (
                    // married
                    LET entityIds = (FOR l IN links_view
                        SEARCH l.type == "relationship"
                        AND l.subType == "married"
                        AND (l.from == org.id OR l.to == org.id)
                        RETURN DISTINCT l.from == org.id ? l.to : l.from)
                    FOR entityId IN entityIds
                    //FILTER LENGTH(entityIds) == 2
                    RETURN entityId
        )
  
LIMIT 10
RETURN { id: org._key, marriedIds: marriedIds }
```

from this query i expect the output

id: person1, marriedIds: [ person2, person3 ]
id: person2, marriedIds: [ person1 ]
id: person3, marriedIds: [ person1 ]
id: person4, marriedIds: [ person5 ]
id: person5, marriedIds: [ person4 ]

Instead, I get the following output (person4 and person5 with wrong married ids)
```
[
  {
    "id": "person1",
    "marriedIds": [
      "person2",
      "person3"
    ]
  },
  {
    "id": "person5",
    "marriedIds": [
      "person1"
    ]
  },
  {
    "id": "person4",
    "marriedIds": [
      "person1"
    ]
  },
  {
    "id": "person3",
    "marriedIds": [
      "person1"
    ]
  },
  {
    "id": "person2",
    "marriedIds": [
      "person1"
    ]
  }
]
```

**Comment by aveiros** [comment]
If i pass a specific org._key, i get the correct results

```// person
FOR org IN entities_view
SEARCH org.type == "person"
AND org._key == "person4"

        // married
        LET marriedIds = (
                    // married
                    LET entityIds = (FOR l IN links_view
                        SEARCH l.type == "relationship"
                        AND l.subType == "married"
                        AND (l.from == org.id OR l.to == org.id)
                        RETURN DISTINCT l.from == org.id ? l.to : l.from)
                    FOR entityId IN entityIds
                    //FILTER LENGTH(entityIds) == 2
                    RETURN entityId
        )
  
LIMIT 10
RETURN { id: org._key, marriedIds: marriedIds }```

**Comment by aveiros** [comment]
Reproduced in 3.4.3

**Comment by gnusi** [comment]
Hi @aveiros ,
Thank you for submitting the issue. We have investigated it and can confirm a bug. Corresponding fix will be available soon in `devel` and become a part of the upcoming minor release `3.4.4`.
We'll keep you updated.

**Comment by gnusi** [comment]
@aveiros Fix is in `devel` and `3.4`. In case if you need any further assistance feel free to let us know.

**Comment by aveiros** [comment]
@gnusi thanks for quickly reacting to this bug.

**Comment by maxkernbach** [comment]
Hi @aveiros,

did you already have a chance to test 3.4.4, which includes a fix for this issue?
The latest release can be found in the [download section](https://www.arangodb.com/download-major/).

**Comment by aveiros** [comment]
@maxkernbach yes I did. It's fixed on 3.4.4.
This can be closed.
